[Feature] Decay functions

FBruzzesi opened the pull request:
# Description

Introduces built-in decay functions as discussed in #581

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected): Fixes #581 


# Checklist:

- [X] My code follows the style guidelines (flake8)
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (also to the readme.md)
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added tests to check whether the new feature adheres to the sklearn convention
- [X] New and existing unit tests pass locally with my changes

FBruzzesi commented on sklego/meta/_decay_utils.py:
```diff
@@ -0,0 +1,272 @@
+    """
+
+    def __init__(self, n_steps=None, step_size=None, min_value=0.0, max_value=1.0):
+        self.n_steps = n_steps
+        self.step_size = step_size
```
Not a huge fan of this implementation, looking for advice 😊

FBruzzesi replied:
Moved validation to `__init__`

FBruzzesi commented on sklego/meta/_decay_utils.py:
```diff
@@ -0,0 +1,272 @@
+            raise ValueError("`growth_rate` must be between 0. and 1.")
+
+        n_samples = y.shape[0]
+        growth_rate = self.growth_rate or 10 / n_samples
```
This tend to be a good default in my experience

FBruzzesi commented on tests/test_meta/test_decay_estimator.py:
```diff
@@ -16,43 +16,59 @@
+@pytest.mark.parametrize("decay_func", ["exponential", "linear", "sigmoid"])
+def test_estimator_checks_regression(test_fn, decay_func):
+    trf = DecayEstimator(LinearRegression(), decay_func=decay_func, check_input=True)
     test_fn(DecayEstimator.__name__, trf)
```
I couldn't manage to change default values in ln22 and ln 29 as well.

FBruzzesi commented on sklego/meta/_decay_utils.py:
```diff
@@ -0,0 +1,272 @@
+import numpy as np
```
Should we add check for `min_value` to be strictly positive? Estimators will handle that anyway, I would say it is redundant

koaning reviewed (commented):
I guess the main thing that's missing here is perhaps a docs page with plots of all of these decay functions?

FBruzzesi commented:
> I guess the main thing that's missing here is perhaps a docs page with plots of all of these decay functions?
> 
> I'll leave it up to @FBruzzesi to determine if it's best to add that later given the big push on the docs. But I'd imagine without those docs people may omit this new feature, which would be a shame.

In order we could:
- Merge doc PRs (whenever ready)
- Add proper user guide/documentation for this feature in this PR
- Finally merge this PR as well

koaning commented:
@FBruzzesi just merged the docs. Feel free to continue on this thread again. I'll try to figure out a nice docs deployment setup from my machine in the meantime.

FBruzzesi commented:
Ready for review!

- Switched from classes to functions to have a unified "api" when passing and calling a custom function
- Added section called **Decay Functions** in the **Meta Models** user guide
- Added API section for the functions as their signature may be useful to anyone who wants to tinker with values
